src/inc/Lfo.h

```diff
@@ -0,0 +1,88 @@
+#if !defined(__Lfo_hdr__)
+#define __Lfo_hdr__
+
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include "ErrorDef.h"
+#include "Synthesis.h"
+#include "RingBuffer.h"
+
+class CLfo
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
+    {
+        m_pCRingBuffer = 0;
+        m_fRead = 0;
+        m_fModFreq = fModFreq;
+        m_fSampleRate = fSampleRate;
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        m_pCRingBuffer = new CRingBuffer<float>(m_iBufferLength);
+    }
+    
+    ~CLfo()
+    {
+        delete m_pCRingBuffer;
+        m_pCRingBuffer = 0;
+    }
+    
+    Error_t setParam(float fParamValue)
+    {
+        m_fModFreq = fParamValue;
+        float *pSine = new float [m_iBufferLength];
+        
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        }
+        else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        delete [] pSine;
+        
+        return Error_t::kNoError;
+    }
+    
+    //return current values
+    float returnLfoVal()
+    {
+        float fLfo = m_pCRingBuffer->get(m_fRead);
+        
+        m_fRead = m_fRead + 1;
+        
+        if (m_fRead >= m_iBufferLength) {
+            m_fRead -= m_iBufferLength;
+        }
+        
+        return fLfo;
+        
+    }
+    
+    void process()
+    {
+        float *pSine = new float [m_iBufferLength];
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
```

Comment:
Is the output computed by sin() function or by different read speed of wavetable?

Reply:
pSine only stores sinewave value in one cycle, but its length would affect the reading speed, as the read index increase one place when input buffer read index increase one place.

Reply:
Now that in the newest version, we get sinewave value in a more clear way.
```
Lfo.h - Line to 60
    Error_t generate () {
        CSynthesis::generateSine(m_pSine, m_fModFreqRatio * m_fSampleRate, m_fSampleRate, m_iWaveLength);
        m_pRingBuffer -> putPostInc(m_pSine, m_iWaveLength);
    }
    
    float getValue () {
        return m_pRingBuffer -> getPostInc();
    }
```